### PR TITLE
Select to edit performance fix

### PR DESCRIPTION
**Pull request by montehurd**
https://phabricator.wikimedia.org/T221608

Fixes variety of small select-to-edit bugs and speeds it up.

### Before:
select - before mov

### After:
select after mov

### CODE REVIEW

File: Wikipedia/assets/codemirror/codemirror-editTextSelection.js [hunk 1]
```diff
@@ -6,92 +6,59 @@ class SelectedAndAdjacentText {
...
+        score: thisMatchScore,
+        matchedWikitextBeforeSelection: {start: 0, end: match.index},
+        matchedWikitextSelection: {start: match.index, end: selectedTextRegex.lastIndex}
+      }    
+    }
   }
-  const match = wikitext.match(regex)
-  if (match === null) {
+
+  if (bestScoredMatch === undefined) {
```
Comment (joewalsh) [inline Wikipedia/assets/codemirror/codemirror-editTextSelection.js]: `null === undefined` is `false` so this fails when `bestScoredMatch` remains `null`

File: Wikipedia/assets/codemirror/codemirror-editTextSelection.js [hunk 2]
```diff
@@ -6,92 +6,59 @@ class SelectedAndAdjacentText {
...
+  return wordsBeforeScore + wordsAfterScore
 }
 
 const wikitextRangeForSelectedAndAdjacentText = (selectedAndAdjacentText, wikitext) => {
-  const regex = selectedAndAdjacentText.regexForLocatingSelectedTextInWikitext(wikitext)
-  if (regex === null) {
-    return null
+  const wordsBeforeFromHTML = getNumberOfWordsFromEndOfString(selectedAndAdjacentText.textBeforeSelectedText, adjacentComparisonWordCount)
+  const wordsAfterFromHTML = getNumberOfWordsFromBeginningOfString(selectedAndAdjacentText.textAfterSelectedText, adjacentComparisonWordCount)
+  const selectedTextRegexPattern = `(${selectedAndAdjacentText.selectedText.replace(/\s+/g, '\\W*(?:(?:[<{][^>]*[}>])*)\\W*')})`
```
Comment (joewalsh) [inline Wikipedia/assets/codemirror/codemirror-editTextSelection.js]: Could the in-between-words regex be updated to be more permissive of text within links? I had some luck with `((\\[\\[[^\\]\\|]+\\|)|{{[^}]*}}|<[^>]*>|\\W)+` but unsure if that has any other consequences. The case I was looking at was selecting "is a grammatically correct sentence in American English" in [this article](https://en.wikipedia.org/wiki/Buffalo_buffalo_Buffalo_buffalo_buffalo_buffalo_Buffalo_buffalo)

Comment (montehurd) [reply Wikipedia/assets/codemirror/codemirror-editTextSelection.js]: ah good catch! will give it a spin!

File: Wikipedia/assets/codemirror/codemirror-editTextSelection.js (hunk 1 above)
Comment (montehurd) [reply Wikipedia/assets/codemirror/codemirror-editTextSelection.js]: doh! 😜

Comment (montehurd) [reply Wikipedia/assets/codemirror/codemirror-editTextSelection.js]: fixed!

File: Wikipedia/assets/codemirror/codemirror-editTextSelection.js (hunk 2 above)
Comment (montehurd) [reply Wikipedia/assets/codemirror/codemirror-editTextSelection.js]: works great! fixed!

**Pull request merged**